fix(ivy): support schemas at runtime

crisbeto opened the pull request:
Accounts for schemas in when validating properties in Ivy.

This PR resolves FW-819.

A couple of notes:
* I had to rework the test slightly, in order to have it fail when we expect it to. The one in master is passing since Ivy's validation runs during the update phase, rather than creation.
* I had to deviate from the design in FW-819 and not add an `enableSchema` instruction, because the schema is part of the `NgModule` scope, however the scope is only assigned to a component once all of the module's declarations have been resolved and some of them can be async. Instead, I opted to have the `schemas` on the component definition.

crisbeto commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2816,7 +2830,14 @@ export function checkView<T>(hostView: LView, component: T) {
+    } else if (ngDevMode) {
+      // We can skip assigning the schemas in production, because
+      // we only validate the against them in dev mode.
+      const componentDef = getComponentDef(component.constructor);
```
Note: this feels a little dirty, but I couldn't think of a better way of getting the component def. I'm open to suggestions on how to do it better.

mhevery replied:
Yes that is dirty. I think the correct answer is that `schemas` should be part of the `TView`.  `hostView[TView].schemas`

AndrewKushnir commented on packages/core/src/render3/state.ts:
```diff
@@ -309,5 +321,6 @@ export function leaveView(newView: LView): void {
       lView[BINDING_INDEX] = tView.bindingStartIndex;
     }
   }
+  currentSchemas = null;
```
Can we assign `null` as a default value when we define `currentSchemas` above?

crisbeto replied:
Done.

mhevery commented on packages/core/src/render3/instructions.ts:
```diff
@@ -1222,6 +1224,18 @@ function elementPropertyInternal<T>(
     element: RElement | RComment, propName: string, tNode: TNode) {
+  const schemas = getCurrentSchemas();
+  const tagName = tNode.tagName;
+  const matchesSchema = schemas && schemas.some(schema => {
```
https://github.com/angular/angular/blob/master/packages/core/src/render3/PERF_NOTES.md#loops please convert `some` to regular `for` loop

mhevery commented on packages/core/src/render3/state.ts:
```diff
@@ -236,6 +237,17 @@ export function setCurrentQueryIndex(value: number): void {
+/** Current schemas of a view that is being processed. */
+let currentSchemas: SchemaMetadata[]|null = null;
+
+export function getCurrentSchemas(): SchemaMetadata[]|null {
```
If we make `schemas` as part of `TView` than we don't need to store it in `state.ts` and this file should be reverted.

crisbeto committed:
fix(ivy): support schemas at runtime

Accounts for schemas in when validating properties in Ivy.

This PR resolves FW-819.

A couple of notes:
* I had to rework the test slightly, in order to have it fail when we expect it to. The one in master is passing since Ivy's validation runs during the update phase, rather than creation.
* I had to deviate from the design in FW-819 and not add an `enableSchema` instruction, because the schema is part of the `NgModule` scope, however the scope is only assigned to a component once all of the module's declarations have been resolved and some of them can be async. Instead, I opted to have the `schemas` on the component definition.

crisbeto commented:
@mhevery I've reworked it based on the feedback. Also I noticed too late that I was basing the changes over my initial commit, rather than in a fixup.

kara commented on packages/core/test/linker/ng_module_integration_spec.ts:
```diff
@@ -266,22 +266,25 @@ function declareTests(config?: {useJit: boolean}) {
+           @NgModule({
+             schemas: [CUSTOM_ELEMENTS_SCHEMA],
+             declarations: [ComponentUsingInvalidProperty],
+             entryComponents: [ComponentUsingInvalidProperty]
```
Seems odd that we are adding this to entry components - can we remove it?

crisbeto replied:
We can't, because we're compiling the component dynamically so that the template function is run. In ViewEngine the check was run when the module was compiled so this wasn't necessary.

kara commented on packages/core/src/render3/instructions.ts:
```diff
@@ -1221,7 +1227,21 @@ function elementPropertyInternal<T>(
+
+  // If there are schemas and the element matches any of the schemas, we shouldn't throw.
+  if (schemas) {
+    for (let i = 0; i < schemas.length; i++) {
```
Nit: What do you think about breaking this for loop out into a function?

```ts
if (matchingSchemas(hostView, tagName)) {
   return;
}
```

crisbeto replied:
Done.

kara reviewed (approved):
Looks good, just minor nits

crisbeto commented:
I've addressed the latest set of feedback @kara.

mhevery commented on packages/compiler/src/compiler_facade_interface.ts:
```diff
@@ -97,6 +97,7 @@ export interface R3NgModuleMetadataFacade {
   imports: Function[];
   exports: Function[];
   emitInline: boolean;
+  schemas?: {name: string}[];
```
Please don't mark it as optional `?`. Making it optional means that both shapes are valid, with and without `schemas` property. This can potentially turn the code megamorphic. I think the correct type is `schemas: {name:string}[]|null`. This way you are saying that the property is required.

mhevery commented on packages/compiler/src/render3/r3_module_compiler.ts:
```diff
@@ -57,21 +57,27 @@ export interface R3NgModuleMetadata {
+  /**
+   * The set of schemas that declare elements to be allowed in the NgModule.
+   */
+  schemas?: R3Reference[]|null;
```
Please don't mark it as optional `?`. Making it optional means that both shapes are valid, with and without `schemas` property. This can potentially turn the code megamorphic. I think the correct type is `schemas: {name:string}[]|null`. This way you are saying that the property is required.

mhevery commented on packages/core/src/compiler/compiler_facade_interface.ts:
```diff
@@ -97,6 +97,7 @@ export interface R3NgModuleMetadataFacade {
   imports: Function[];
   exports: Function[];
   emitInline: boolean;
+  schemas?: {name: string}[];
```
Please don't mark it as optional `?`. Making it optional means that both shapes are valid, with and without `schemas` property. This can potentially turn the code megamorphic. I think the correct type is `schemas: {name:string}[]|null`. This way you are saying that the property is required.

mhevery commented on packages/core/src/metadata/ng_module.ts:
```diff
@@ -28,6 +28,7 @@ import {TypeDecorator, makeDecorator} from '../util/decorators';
 export interface NgModuleTransitiveScopes {
   compilation: {directives: Set<any>; pipes: Set<any>;};
   exported: {directives: Set<any>; pipes: Set<any>;};
+  schemas?: SchemaMetadata[]|null;
```
Please don't mark it as optional `?`. Making it optional means that both shapes are valid, with and without `schemas` property. This can potentially turn the code megamorphic. I think the correct type is `schemas: {name:string}[]|null`. This way you are saying that the property is required.

mhevery commented on packages/core/src/metadata/ng_module.ts:
```diff
@@ -67,6 +68,9 @@ export interface NgModuleDef<T> {
   transitiveCompileScopes: NgModuleTransitiveScopes|null;
+
+  /** The set of schemas that declare elements to be allowed in the NgModule. */
+  schemas?: SchemaMetadata[]|null;
```
Please don't mark it as optional `?`. Making it optional means that both shapes are valid, with and without `schemas` property. This can potentially turn the code megamorphic. I think the correct type is `schemas: {name:string}[]|null`. This way you are saying that the property is required.

mhevery commented on packages/core/src/render3/instructions.ts:
```diff
@@ -722,12 +723,14 @@ function saveResolvedLocalsInData(
     templateFn: ComponentTemplate<any>, consts: number, vars: number,
     directives: DirectiveDefListOrFactory | null, pipes: PipeDefListOrFactory | null,
-    viewQuery: ComponentQuery<any>| null): TView {
+    viewQuery: ComponentQuery<any>| null, schemas?: SchemaMetadata[] | null): TView {
```
for performance it is better to have the same number of arguments. Since this is private API I would make `schemas` required and caller will have to pass in `null`

mhevery commented on packages/core/src/render3/instructions.ts:
```diff
@@ -748,11 +751,13 @@ export function getOrCreateTView(
     viewIndex: number, templateFn: ComponentTemplate<any>| null, consts: number, vars: number,
     directives: DirectiveDefListOrFactory | null, pipes: PipeDefListOrFactory | null,
-    viewQuery: ComponentQuery<any>| null): TView {
+    viewQuery: ComponentQuery<any>| null, schemas?: SchemaMetadata[] | null): TView {
```
same here

mhevery commented on packages/core/src/render3/interfaces/definition.ts:
```diff
@@ -255,6 +254,11 @@ export interface ComponentDef<T> extends DirectiveDef<T> {
+  /**
+   * The set of schemas that declare elements to be allowed in the component's template.
+   */
+  schemas?: SchemaMetadata[]|null;
```
Please don't mark it as optional `?`. Making it optional means that both shapes are valid, with and without `schemas` property. This can potentially turn the code megamorphic. I think the correct type is `schemas: {name:string}[]|null`. This way you are saying that the property is required.

mhevery commented on packages/core/src/render3/interfaces/view.ts:
```diff
@@ -528,6 +529,11 @@ export interface TView {
+  /**
+   * Set of schemas that declare elements to be allowed inside the view.
+   */
+  schemas?: SchemaMetadata[]|null;
```
Please don't mark it as optional `?`. Making it optional means that both shapes are valid, with and without `schemas` property. This can potentially turn the code megamorphic. I think the correct type is `schemas: {name:string}[]|null`. This way you are saying that the property is required.

mhevery commented on packages/core/src/render3/instructions.ts:
```diff
@@ -1234,6 +1245,22 @@ function validateAgainstUnknownProperties(
+function matchingSchemas(hostView: LView, tagName: string | null): boolean {
+  const schemas = hostView[TVIEW].schemas;
+
+  if (schemas) {
```
use `schemas !== null` as it is slightly faster.

mhevery reviewed (changes requested):
Left some more comments. You may consider reading these articles:
- https://mrale.ph/blog/2015/01/11/whats-up-with-monomorphism.html
- https://richardartoul.github.io/jekyll/update/2015/04/26/hidden-classes.html

JoostK commented on packages/core/test/linker/ng_module_integration_spec.ts:
```diff
@@ -275,6 +275,10 @@ function declareTests(config?: {useJit: boolean}) {
+
+             // Note that we need to add the component to `entryComponents`, because of the
+             // `createComp` call below. In Ivy the property validation happens during the
+             //  update phase so we need to create the component, in order for it to run.
```
Either my understanding of Ivy is broken, or this shouldn't be the case. Basically, `entryComponents` no longer has value with Ivy, if I'm not mistaken. All that entry components achieve is a reference to a component factory being held in a component factory resolver, such that a reference can be obtained to a component factory even when a component is not statically referenced from within a template. With Ivy a component itself holds its definition object, so if you have a reference to the component you can let Angular render it.

Since we do have a reference to `ComponentUsingInvalidProperty`, we should be able to just render that. It should have been compiled regardless of it being an entry point or not.

I am not requesting you change this, but I am just wondering if my understanding of entry components is perhaps incorrect (if you have observed that it does in fact make a difference, I wouldn't be able to explain why it doesn't work without this).

crisbeto replied:
Note that this test also runs against ViewEngine.

JoostK replied:
@crisbeto Ah, I see :) I interpreted the comment differently, as if the inclusion as entry component is necessary for Ivy. It's the other way around. Thanks for enlightening me 👍

mhevery commented:
http://test/OCL:233687876:BASE:233687884:1550036717498:eef4a766

JoostK commented on packages/core/src/render3/definition.ts:
```diff
@@ -237,6 +237,11 @@ export function defineComponent<T>(componentDefinition: {
+  /**
+   * The set of schemas that declare elements to be allowed in the component's template.
+   */
+  schemas: SchemaMetadata[] | null;
```
The `defineComponent` arg is not expected to be monomorphic, so this one should probably be optional. I was triggered by the many test changes that now pass in `schemas: null`. Doing so would require the compiler to generate additional code, which is undesirable.

crisbeto replied:
That's what I had initially, but I've changed it as a result of https://github.com/angular/angular/pull/28637#discussion_r256096009.

JoostK replied:
@crisbeto I think that is a mistake. Let's check with @mhevery

mhevery commented:
This PR breaks g3

```
/build/work/f4ee8dd3fa0818fc4f56ff9a63ed8456/google3/third_party/javascript/angular2/rc/packages/core/src/render3/instructions.ts -> /build/work/f4ee8dd3fa0818fc4f56ff9a63ed8456/google3/third_party/javascript/angular2/rc/packages/core/public_api.ts -> /build/work/f4ee8dd3fa0818fc4f56ff9a63ed8456/google3/third_party/javascript/angular2/rc/packages/core/src/core.ts -> /build/work/f4ee8dd3fa0818fc4f56ff9a63ed8456/google3/third_party/javascript/angular2/rc/packages/core/src/metadata.ts -> /build/work/f4ee8dd3fa0818fc4f56ff9a63ed8456/google3/third_party/javascript/angular2/rc/packages/core/src/metadata/directives.ts -> /build/work/f4ee8dd3fa0818fc4f56ff9a63ed8456/google3/third_party/javascript/angular2/rc/packages/core/src/render3/jit/directive.ts -> /build/work/f4ee8dd3fa0818fc4f56ff9a63ed8456/google3/third_party/javascript/angular2/rc/packages/core/src/render3/jit/environment.ts -> /build/work/f4ee8dd3fa0818fc4f56ff9a63ed8456/google3/third_party/javascript/angular2/rc/packages/core/src/render3/index.ts -> /build/work/f4ee8dd3fa0818fc4f56ff9a63ed8456/google3/third_party/javascript/angular2/rc/packages/core/src/render3/component.ts -> /build/work/f4ee8dd3fa0818fc4f56ff9a63ed8456/google3/third_party/javascript/angular2/rc/packages/core/src/render3/global_utils.ts -> /build/work/f4ee8dd3fa0818fc4f56ff9a63ed8456/google3/third_party/javascript/angular2/rc/packages/core/src/render3/global_utils_api.ts -> /build/work/f4ee8dd3fa0818fc4f56ff9a63ed8456/google3/third_party/javascript/angular2/rc/packages/core/src/render3/instructions.ts
```

crisbeto committed:
fixup! fix(ivy): support schemas at runtime

mhevery commented:
http://test/OCL:233687876:BASE:233803500:1550088644497:8f7fa82f
